src/eureka/S4_generate_lightcurves/plots_s4.py

```diff
@@ -331,6 +318,99 @@ def lc_driftcorr(meta, wave_1d, optspec_in, optmask=None, scandir=None):
     return
 
 
+def mad_outliers(meta, lc, spec):
+    '''Plot spectroscopic MAD values to identify outliers. (Figs 4106)
+    Outliers can be masked using `mask_columns` in the Stage 4 ECF.
+
+    Parameters
+    ----------
+    meta : eureka.lib.readECF.MetaClass
+        The metadata object.
+    lc : Xarray Dataset
+        The Dataset object containing binned light curve and time data.
+    spec : Xarray Dataset
+        The Dataset object containing spectroscopic LC and time data.
+    '''
+    # Normalize the light curve
+    wave_1d = spec.wave_1d.values
+    iwmin = np.nanargmin(np.abs(wave_1d - meta.wave_min))
+    iwmax = np.nanargmin(np.abs(wave_1d - meta.wave_max))
+    optspec = spec.optspec.values[:, iwmin:iwmax]
+    opterr = spec.opterr.values[:, iwmin:iwmax]
+    optmask = spec.optmask.values[:, iwmin:iwmax]
+    norm_lcdata, norm_lcerr = util.normalize_spectrum(meta, optspec, opterr,
+                                                      optmask=optmask)
+    norm_lcdata = norm_lcdata.filled(np.nan)
+    norm_lcerr = norm_lcerr.filled(np.nan)
+
+    # Compute unbinned LC MAD values, then scale
+    numx = norm_lcdata.shape[1]
+    mad = np.zeros(numx)
+    for ii in range(numx):
+        mad[ii] = util.get_mad_1d(norm_lcdata[:, ii])
+
+    # Compute mean abs deviation from "white" LC, then scale
+    optspec_mean = np.nanmean(norm_lcdata, axis=1)
+    dev = np.zeros(numx)
+    for ii in range(numx):
+        dev[ii] = np.ma.mean(np.ma.abs((norm_lcdata[:, ii] - optspec_mean)))
+    dev /= np.nanmean(dev)/np.nanmean(mad)
+
+    # Remove broad trends from native-resolution MAD values
+    mask = np.isnan(mad)
+    x = spec.x[iwmin:iwmax]
+    x_mask = x[~mask]
+    smoothed_mad = smooth.medfilt(mad[~mask], window_len=meta.box_width)
+    residual_mad = mad[~mask] - smoothed_mad
+    smoothed_dev = smooth.medfilt(dev[~mask], window_len=meta.box_width)
+    residual_dev = dev[~mask] - smoothed_dev
+
+    # Identify only high outliers from residuals
+    masked_mad = sigma_clip(residual_mad, sigma_upper=meta.mad_sigma,
+                            sigma_lower=100, maxiters=meta.maxiters,
+                            masked=True, copy=True)
+    masked_dev = sigma_clip(residual_dev, sigma_upper=meta.mad_sigma,
+                            sigma_lower=100, maxiters=meta.maxiters,
+                            masked=True, copy=True)
+    x_mad_outliers = x_mask[np.where(masked_mad.mask)[0]]
+    x_dev_outliers = x_mask[np.where(masked_dev.mask)[0]]
+    outliers = np.union1d(x_mad_outliers, x_dev_outliers)
```

Comment:
These calculations should be moved outside of the plotting function as the MAD clipping shouldn't depend on whether or not the plot was made, and this code is not purely related to plotting